src/query/block/lazy.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"log"
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+)
+
+// TimeTransform transforms a timestamp
+type TimeTransform func(time.Time) time.Time
+
+// MetaTranform transforms meta data
+type MetaTranform func(meta Metadata) Metadata
+
+// ValueTransform transform a float64
+type ValueTransform func(float64) float64
+
+// LazyOpts are used to create a lazy block
```

Comment:
Rather than doing it like this, can we refactor LazyOpts be an interface [like this one](https://sourcegraph.com/github.com/m3db/m3/-/blob/src/query/models/options.go#L85:12), and have the transforms default to a noop function?